ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/KeyValueParticipantStateReader.scala

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.api
+
+import akka.NotUsed
+import akka.stream.Materializer
+import akka.stream.scaladsl.Source
+import com.daml.ledger.participant.state.kvutils.{Envelope, KeyValueConsumption}
+import com.daml.ledger.participant.state.v1._
+import com.digitalasset.daml.lf.data.Time
+import com.digitalasset.ledger.api.health.{HealthStatus, Healthy}
+
+class KeyValueParticipantStateReader(reader: LedgerReader)(implicit materializer: Materializer)
+    extends ReadService {
+  override def getLedgerInitialConditions(): Source[LedgerInitialConditions, NotUsed] =
+    Source.single(createLedgerInitialConditions())
+
+  override def stateUpdates(beginAfter: Option[Offset]): Source[(Offset, Update), NotUsed] =
+    reader
+      .events(toReaderOffset(beginAfter))
+      .flatMapConcat { record =>
+        val updates = Envelope.open(record.envelope) match {
+          case Right(Envelope.LogEntryMessage(logEntry)) =>
+            KeyValueConsumption
+              .logEntryToUpdate(record.entryId, logEntry)
+              .zipWithIndex
+              .map {
+                case (entry, index) =>
+                  (toReturnedOffset(index, record.offset), entry)
+              }
+          case _ => Seq.empty
+        }
+        Source.fromIterator(() => updates.iterator)
+      }
+      .filter {
+        case (offset, _) => beginAfter.forall(offset > _)
```

Comment:
Why do we need to filter again? Shouldn't this be already handled by line 21?

Reply:
This needs to be done as the `LedgerReader` implementation may not support filtering.